src/lib.rs

```diff
@@ -20,15 +41,47 @@ pub enum IoAddress {
     Mmio(GuestAddress),
 }
 
+impl IoAddress {
+    // Get the raw value of IO Address to make operation simple.
+    fn raw_value(&self) -> u64 {
+        match *self {
+            IoAddress::Pio(p) => u64::from(p),
+            IoAddress::Mmio(m) => m.raw_value(),
+        }
+    }
+}
+
+impl Eq for IoAddress {}
+
+impl PartialEq for IoAddress {
+    fn eq(&self, other: &IoAddress) -> bool {
+        self.raw_value() == other.raw_value()
+    }
+}
+
+impl Ord for IoAddress {
+    fn cmp(&self, other: &IoAddress) -> Ordering {
+        self.raw_value().cmp(&other.raw_value())
+    }
+}
+
+impl PartialOrd for IoAddress {
+    fn partial_cmp(&self, other: &IoAddress) -> Option<Ordering> {
+        self.raw_value().partial_cmp(&other.raw_value())
+    }
+}
+
 /// Device IO trait.
 /// A device supporting memory based I/O should implement this trait, then
 /// register itself against the different IO type ranges it handles.
 /// The VMM will then dispatch IO (PIO or MMIO) VM exits by calling into the
 /// registered devices read or write method from this trait.
+/// The DeviceIo trait adopts the interior mutability pattern
```

Comment:
Having interior mutability here is a thorny issue and needs a separate discussion :( Why do you favor this over the previous interface?

Reply:
Some users of firecracker have mentioned that lacking support of multi-queue causes performance penalty, so we hope we could enable mutli-queue for virtio devices.
Moving lock into the device backend driver is the first step:)

Reply:
> Having interior mutability here is a thorny issue and needs a separate discussion :( Why do you favor this over the previous interface?

@alexandruag Any further comments on this? I'm ok with this PR, but I'd like to get your opinion on it and see if you want to block it until we agree on the inner mutability part.

Reply:
Hi, apologies for the delay. The problem with assuming interior mutability is many devices mutate their state during normal operation, but have to go out of their way to actually implement interior mutability. I don't know what the best approach is long term (hopefully it'll become clearer as we start putting things together), but one potential solution right now is to add another trait (something like `DeviceIoMut`) that defines the same methods as `DeviceIo` but requiring `&mut self` instead of `&self`, and then also add a blanket implementation such as `impl<T> DeviceIo for Mutex<T> where T: DeviceIoMut { ... }`, that allows a `Mutex<T>` to be used as a `DeviceIo` object automatically as long as the inner type implements `DeviceIoMut` (same goes for other standard wrappers that offer interior mutability like `RefCell`). Does that make any sense?

Reply:
Great idea, this satisfies both side with an zero overhead abstraction.
How about merging the PR first and sending another PR for the DeviceIoMut proposal?